### ISSUE TITLE
MKL package intel-mkl-64bit-2020.0-088 installation FAILED

### ISSUE THREAD

**Issue by DanielEgoli** [kaldi-asr/kaldi#4366]
When trying to install the install_mkl.sh file I receive this error:

+ exec sudo extras/install_mkl.sh -sp debian intel-mkl-64bit-2020.0-088
+ apt-get update
Hit:1 http://us-east-1.ec2.archive.ubuntu.com/ubuntu bionic InRelease
Hit:2 http://us-east-1.ec2.archive.ubuntu.com/ubuntu bionic-updates InRelease                                        
Hit:3 http://us-east-1.ec2.archive.ubuntu.com/ubuntu bionic-backports InRelease                                      
Hit:4 https://nvidia.github.io/libnvidia-container/stable/ubuntu18.04/amd64  InRelease                               
Hit:5 https://nvidia.github.io/nvidia-container-runtime/ubuntu18.04/amd64  InRelease                                 
Hit:6 https://nvidia.github.io/nvidia-docker/ubuntu18.04/amd64  InRelease                                            
Hit:7 https://apt.repos.neuron.amazonaws.com bionic InRelease                                                        
Hit:8 https://download.docker.com/linux/ubuntu bionic InRelease                                                      
Hit:9 http://security.ubuntu.com/ubuntu bionic-security InRelease                                                    
Hit:10 http://ppa.launchpad.net/openjdk-r/ppa/ubuntu bionic InRelease    
Traceback (most recent call last):                 
  File "/usr/lib/cnf-update-db", line 8, in <module>
    from CommandNotFound.db.creator import DbCreator
  File "/usr/lib/python3/dist-packages/CommandNotFound/db/creator.py", line 11, in <module>
    import apt_pkg
ModuleNotFoundError: No module named 'apt_pkg'
Reading package lists... Done
E: Problem executing scripts APT::Update::Post-Invoke-Success 'if /usr/bin/test -w /var/lib/command-not-found/ -a -e /usr/lib/cnf-update-db; then /usr/lib/cnf-update-db > /dev/null; fi'
E: Sub-process returned an error code
extras/install_mkl.sh: MKL package intel-mkl-64bit-2020.0-088 installation FAILED.

I cannot progress any further with installation - can you fix?

**Comment by jtrmal** [comment]
to me it looks more like some deeper system issue -- no better idea
y.

**Comment by danpovey** [comment]
I had that same issue with apt_pkg, it's a problem in Ubuntu when you
update Python.  It's clearly a very bad design issue in Ubuntu, that
the system package manager (which absolutely needs to be robust) is so
easily broken by changes in installed packages.  I really don't
know what they could have been thinking when they decided to make apt
dependent on python packages in that way.

I remember it was very difficult to fix but can't share much more detail
than that.  I think one of the fixes below


https://stackoverflow.com/questions/13708180/python-dev-installation-error-importerror-no-module-named-apt-pkg

worked for me, I think it involved force-deleting packages and reinstalling
them.

**Comment by danpovey** [comment]
it's possible that apt-get update is not really necessary in that mkl thing.
Also (@kkm)  last time I needed to install MKL on ubuntu, I found that locating the system package with `apt-cache search mkl` and installing it was easier than using Kaldi's own script.

**Comment by danpovey** [comment]
I mean @kkm000

**Comment by kkm000** [comment]
@danpovey, this was done in #3433.

Ubuntu is generally unstable. They take Debian-stable, and add packages from the next two versions (testing and unstable, in Debian lingo). Debian term for this is [Frankendebian](https://wiki.debian.org/DontBreakDebian#Don.27t_make_a_FrankenDebian). I used to run both Debian stable and testing as my desktop and in production, and had much fewer problems that I did with Ubuntu. I do not know why Ubuntu is so popular that people run it _in production_, my apt! That's certainly asking for a trouble. Personally, I have always suggested pinning down all apt feeds except security down to 100, but that's an advanced technique.

The thing called just "intel-mkl-something" or how it was called is going to change randomly, as everything else does in the Ubuntu world. I could not even figure out which version of MKL it was, 2020.1 or 2020.3. Unless you update Ubuntu daily, by the time you get to it the list of upgradable packages runs into hundreds; it's a good idea to review what you're installing, but ain't anyone got time for this. This is why I am totally fine with _supporting_ Ubuntu-repackaged MKL, but would rather avoid _recommending_ it.

@DanielEgoli: The likely fix for your apt breakage is `sudo apt install --reinstall python3-apt`. First make sure `sudo apt update` ends w/o errors. This is not related to the MKL repo.
